sink.js

```diff
@@ -72,19 +72,21 @@ module.exports = writable => async source => {
         await waitForDrainOrClose()
       }
     }
-  } catch (err) {
-    // The writable did not error, give it the error
-    writable.destroy(err)
-  }
 
-  try {
     // Everything is good and we're done writing, end everything
     if (!error && writable.writable) {
       writable.end()
     }
 
     // Wait until we close or finish. This supports halfClosed streams
     await waitForDone()
+  } catch (err) {
+    // the source threw, destroy the stream and throw the error
+    // call end too as we want it to be marked `writable: false`
+    writable.end()
+    writable.destroy()
```

Comment:
PassThrough streams (at least) do not set `.writable` to false when destroyed which seems quite naughty as the node docs say about `.writable`:

> Is true if it is safe to call `writable.write()`, which means the stream has not been destroyed, errored or ended

yet also about `.destroy([error])`:

> After this call, the writable stream has ended and subsequent calls to write() or end() will result in an ERR_STREAM_DESTROYED error.

So you'd think `.writable` would be false after `.destroy([error])` to avoid a `ERR_STREAM_DESTROYED` error but it is not:

```javascript
const { PassThrough } = require('stream')

const p = new PassThrough()

console.info(p.writable) // true
console.info(p.destroyed) // false

p.destroy()

console.info(p.writable) // true
console.info(p.destroyed) // true

p.end()

console.info(p.writable) // false
console.info(p.destroyed) // true
```

Reply:
The reason there are separate try catch statements here is because a signular try/catch does not guarantee the source was the thing that errored. It's possible `waitForDone` can throw from the writable. Leaving this as 2 statements and changing the first catch to:
```
    writable.end()
    writable.destroy()
    error = err
```
would replicate what you're trying to do. However, it would be ideal to have a test that replicates the actual problem to verify this is working.

Reply:
Looks like `waitForDrainOrClose` can also cause a throw from an error on the writable, and that's caught by the try/catch as if it were from the source too.

Reply:
I've updated this to restore the second try/catch